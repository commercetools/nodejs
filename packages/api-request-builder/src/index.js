// eslint-disable-next-line import/prefer-default-export
export { default as createRequestBuilder } from './create-request-builder'
export * as features from './features'
