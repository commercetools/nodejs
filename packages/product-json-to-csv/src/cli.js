import fs from 'fs'
import readline from 'readline'
import { getCredentials } from '@commercetools/get-credentials'
import pino from 'pino'
import PrettyError from 'pretty-error'
import yargs from 'yargs'

import ProductJsonToCsv from './main'
import { description } from '../package.json'

process.title = 'product-json-to-csv'

const args = yargs
  .usage(
    `
Usage: $0 [options]
${description}`
  )
  .showHelpOnFail(false)
  .option('projectKey', {
    alias: 'p',
    describe: 'API project key',
    demandOption: true,
    type: 'string',
  })
  .option('apiUrl', {
    default: 'https://api.commercetools.com',
    describe: 'The host URL of the HTTP API service',
    type: 'string',
  })
  .option('authUrl', {
    default: 'https://auth.commercetools.com',
    describe: 'The host URL of the OAuth API service',
    type: 'string',
  })
  .option('accessToken', {
    describe: `CTP client access token
Required scopes: ['view_products']`,
    type: 'string',
  })
  .option('template', {
    alias: 't',
    describe:
      'CSV file containing your header that defines what you want to export.',
  })
  .coerce('template', arg => {
    if (fs.existsSync(arg)) {
      if (arg.match(/\.csv$/i)) return fs.createReadStream(String(arg))

      throw new Error('Invalid file format. Must be CSV file')
    }
    throw new Error('File cannot be reached or does not exist')
  })
  .option('input', {
    alias: 'i',
    default: 'stdin',
    describe: 'Path from which to read product chunks.',
  })
  .coerce('input', arg => {
    if (arg === 'stdin') return process.stdin

    if (fs.existsSync(arg)) {
      if (arg.match(/\.json$/i)) return fs.createReadStream(String(arg))

      throw new Error('Invalid input file format. Must be JSON file')
    }
    throw new Error('Input file cannot be reached or does not exist')
  })
  .option('output', {
    alias: 'o',
    default: 'stdout',
    describe: 'Path to output',
    type: 'string',
  })
  .coerce('output', arg => {
    if (arg !== 'stdout') return fs.createWriteStream(String(arg))

    return process.stdout
  })
  .option('referenceCategoryBy', {
    choices: ['name', 'key', 'externalId', 'namedPath'],
    default: 'name',
    describe:
      'Define which identifier should be used for the categories column.',
  })
  .option('referenceCategoryOrderHintBy', {
    choices: ['name', 'key', 'externalId', 'namedPath'],
    default: 'name',
    describe:
      'Define which identifier should be used for the categoryOrderHints column.',
  })
  .option('fillAllRows', {
    describe:
      'Define if product attributes like name should be added to each variant row.',
    type: 'boolean',
  })
  .option('onlyMasterVariants', {
    describe: 'Export only masterVariants from products.',
    type: 'boolean',
    default: false,
  })
  .option('language', {
    alias: 'l',
    default: 'en',
    describe:
      'Language(s) used for localised attributes such as category names.' +
      ' Can contain multiple languages delimited by comma ","',
    type: 'string',
  })
  .option('delimiter', {
    alias: 'd',
    default: ',',
    describe: 'Used CSV delimiter.',
  })
  .option('multiValueDelimiter', {
    alias: 'm',
    default: ';',
    describe: 'Used CSV delimiter in multiValue fields.',
  })
  .option('encoding', {
    alias: 'e',
    default: 'utf8',
    describe: 'Encoding used when saving data to output file',
    type: 'string',
  })
  .option('logLevel', {
    default: 'info',
    describe: 'Logging level: error, warn, info or debug',
    type: 'string',
  })
  .option('prettyLogs ', {
    describe: 'Pretty print logs to the terminal',
    type: 'boolean',
  })
  .option('logFile', {
    default: 'product-json-to-csv.log',
    describe: 'Path to file where logs should be saved',
    type: 'string',
  }).argv

// instantiate logger
const loggerConfig = {
  level: args.logLevel,
  prettyPrint: args.prettyLogs,
}

// If the stdout is used for a data output, save all logs to a log file.
// pino writes logs to stdout by default
let logDestination
if (args.output === process.stdout)
  logDestination = fs.createWriteStream(args.logFile)

const logger = pino(loggerConfig, logDestination)

// print errors to stderr if we use stdout for data output
// if we save data to output file errors are already logged by pino
const logError = error => {
  const errorFormatter = new PrettyError()

  if (args.logLevel === 'debug')
    process.stderr.write(`ERR: ${errorFormatter.render(error)}`)
  else process.stderr.write(`ERR: ${error.message || error}`)
}

const errorHandler = errors => {
  if (Array.isArray(errors)) errors.forEach(logError)
  else logError(errors)

  process.exitCode = 1
}

// Retrieve the headers from the template file
// Only the first line of the file is read
const getHeaders = _args =>
  new Promise((resolve, reject) => {
    if (!_args.template) resolve(null)
    else {
      const rl = readline.createInterface({
        input: _args.template,
      })
      rl.on('error', reject)
      rl.on('line', line => {
        rl.close()
        resolve(line.split(_args.delimiter))
      })
    }
  })

const resolveCredentials = _args => {
  if (_args.accessToken) return Promise.resolve({})
  return getCredentials(_args.projectKey)
}

// Register error listener
args.output.on('error', errorHandler)

Promise.all([getHeaders(args), resolveCredentials(args)])
  .then(([headerFields, credentials]) => {
    const apiConfig = {
      host: args.authUrl,
      apiUrl: args.apiUrl,
      projectKey: args.projectKey,
      credentials,
    }
    const productJsonToCsvConfigOptions = {
      categoryBy: args.referenceCategoryBy,
      categoryOrderHintBy: args.referenceCategoryOrderHintBy,
      delimiter: args.delimiter,
      fillAllRows: args.fillAllRows,
      language: args.language.split(',')[0], // take first language as primary
      languages: args.language.split(','), // allow export to multiple languages
      multiValueDelimiter: args.multiValueDelimiter,
      onlyMasterVariants: args.onlyMasterVariants,
      headerFields,
    }
    const parserLogger = {
      error: logger.error.bind(logger),
      warn: logger.warn.bind(logger),
      info: logger.info.bind(logger),
      debug: logger.debug.bind(logger),
    }
    const accessToken = args.accessToken

    return new ProductJsonToCsv(
      apiConfig,
      productJsonToCsvConfigOptions,
      parserLogger,
      accessToken
    )
  })
  .then(productJsonToCsv => {
    args.input.setEncoding(args.encoding)
    productJsonToCsv.run(args.input, args.output)
  })
  .catch(errorHandler)
