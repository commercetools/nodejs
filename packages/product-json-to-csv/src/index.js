export { default } from './main'
export { default as MapProductData } from './map-product-data'
export { default as mapHeaders } from './map-headers'
export {
  onStreamsFinished,
  archiveDir,
  writeToSingleCsvFile,
  writeToZipFile,
} from './writer'
