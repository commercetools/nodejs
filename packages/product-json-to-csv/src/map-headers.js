/* @flow */

type HeaderObj = {
  label: string,
  value: string,
}

export default function mapHeaders(headers: Array<string>): Array<HeaderObj> {
  return headers.map(
    (header: string): HeaderObj => {
      if (header === '_published' || header === 'published')
        return {
          label: '_published',
          value: 'published',
        }
      if (header === '_hasStagedChanges' || header === 'hasStagedChanges')
        return {
          label: '_hasStagedChanges',
          value: 'hasStagedChanges',
        }
      if (header === 'variantId' || header === 'variant.id')
        return {
          label: 'variantId',
          value: 'variant.id',
        }
      if (header === 'variantKey' || header === 'variant.key')
        return {
          label: 'variantKey',
          value: 'variant.key',
        }
      if (header === 'sku' || header === 'variant.sku')
        return {
          label: 'sku',
          value: 'variant.sku',
        }
      if (header === 'tax' || header === 'taxCategory')
        return {
          label: 'tax',
          value: 'taxCategory',
        }
      if (header === 'images' || header === 'variant.images')
        return {
          label: 'images',
          value: 'variant.images',
        }
      return {
        label: header,
        value: header,
      }
    }
  )
}
