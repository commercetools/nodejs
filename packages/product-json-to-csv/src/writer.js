import archiver from 'archiver'
import EmitOnce from 'single-emit'
import fs from 'fs'
import json2csv from 'json2csv'
import path from 'path'
import slugify from 'slugify'
import tmp from 'tmp'
import mapHeaders from './map-headers'

// Accept a highland stream and write the output to a single file
export function writeToSingleCsvFile(
  productStream,
  output,
  logger,
  headerFields,
  del
) {
  const trimmedHeaders = headerFields.map(header => header.trim())
  output.write(`${trimmedHeaders.join(del)}\n`) // Write headers first
  const columnNames = mapHeaders(trimmedHeaders)
  productStream
    .each(product => {
      const csvData = json2csv({
        data: product,
        fields: columnNames,
        hasCSVColumnTitle: false,
        quotes: '',
        del,
      })
      output.write(`${csvData}\n`)
    })
    .done(() => {
      if (output !== process.stdout) output.end()

      logger.info('All products have been written to CSV file')
    })
}

// Accept a highland stream and write the output to multiple files per
// product type, then compress all files to a zip file
export function writeToZipFile(productStream, output, logger, del) {
  const tmpDir = tmp.dirSync({ unsafeCleanup: true }).name
  tmp.setGracefulCleanup()
  let currentProductType
  let fileStream
  let columnNames
  const columnNamesCache = {}
  const streamCache = {}
  productStream
    .each(product => {
      let hasCSVColumnTitle = false
      // Process this block only if item is a masterVariant and was
      // not the last processed item
      if (product.productType && product.productType !== currentProductType) {
        // get product columnNames from cache or masterVariant.
        // variants may not have all header fields so should not be relied on
        if (columnNamesCache[product.productType])
          columnNames = columnNamesCache[product.productType]
        else {
          hasCSVColumnTitle = true
          columnNames = mapHeaders(Object.keys(product))
          columnNamesCache[product.productType] = columnNames
        }
        currentProductType = product.productType
        const fileName = `${slugify(product.productType)}.csv`
        const filePath = path.join(tmpDir, fileName)
        // use a stream cache for switching back and forth between file
        // streams and reusing the same write stream
        if (streamCache[currentProductType])
          fileStream = streamCache[currentProductType]
        else {
          fileStream = fs.createWriteStream(filePath)
          streamCache[currentProductType] = fileStream
        }
      }
      const csvData = json2csv({
        data: product,
        fields: columnNames,
        hasCSVColumnTitle,
        quotes: '',
        del,
      })
      fileStream.write(`${csvData}\n`)
    })
    .done(() => {
      const emitOnce = new EmitOnce(streamCache, 'finish')
      const archive = archiver('zip')
      archive.on('error', err => {
        logger.error(err)
        output.emit('error', err)
      })
      emitOnce.on('error', err => {
        logger.error(err)
        output.emit('error', err)
      })
      // close all open file streams
      const streams = Object.keys(streamCache)
      streams.forEach(key => {
        streamCache[key].end()
      })
      // zip files when all file writes have completed
      emitOnce.on('finish', () => {
        archive.pipe(output)
        archive.directory(tmpDir, 'products')
        archive.finalize()
        logger.info('All products have been written to ZIP file')
      })
    })
}
