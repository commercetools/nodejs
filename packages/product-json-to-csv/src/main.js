/* @flow */
import type {
  ApiConfigOptions,
  Category,
  LoggerOptions,
  ParserConfigOptions,
  ProductProjection,
  ProductType,
  ResolvedProductProjection,
  State,
  TaxCategory,
  TypeReference,
} from 'types/product'
import type { Client, ClientRequest, SuccessResult } from 'types/sdk'

import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent'
import highland from 'highland'
import fetch from 'node-fetch'
import Promise from 'bluebird'
import JSONStream from 'JSONStream'
import { memoize } from 'lodash'
import ProductMapping from './map-product-data'
import { writeToSingleCsvFile, writeToZipFile } from './writer'
import pkg from '../package.json'

export default class ProductJsonToCsv {
  // Set flowtype annotations
  accessToken: string

  apiConfig: ApiConfigOptions

  categoriesCache: Object

  client: Client

  parserConfig: ParserConfigOptions

  logger: LoggerOptions

  fetchReferences: Function

  _resolveReferences: Function

  _productMapping: Object

  constructor(
    apiConfig: ApiConfigOptions,
    parserConfig: ParserConfigOptions,
    logger: LoggerOptions,
    accessToken: string
  ) {
    this.apiConfig = apiConfig
    this.client = createClient({
      middlewares: [
        createAuthMiddlewareForClientCredentialsFlow({
          ...this.apiConfig,
          fetch,
        }),
        createUserAgentMiddleware({
          libraryName: pkg.name,
          libraryVersion: pkg.version,
        }),
        createHttpMiddleware({
          host: this.apiConfig.apiUrl,
          fetch,
        }),
      ],
    })

    const defaultConfig = {
      categoryBy: 'name', // key, externalId or namedPath supported
      categoryOrderHintBy: 'name', // key, externalId or name supported
      delimiter: ',',
      fillAllRows: false,
      language: 'en',
      multiValueDelimiter: ';',
    }

    this.parserConfig = { ...defaultConfig, ...parserConfig }
    this.logger = {
      error: () => {},
      warn: () => {},
      info: () => {},
      debug: () => {},
      ...logger,
    }
    this.categoriesCache = {}
    this.accessToken = accessToken

    const mappingParams = {
      fillAllRows: this.parserConfig.fillAllRows,
      categoryBy: this.parserConfig.categoryBy,
      lang: this.parserConfig.language,
      multiValueDelimiter: this.parserConfig.multiValueDelimiter,
    }
    this._productMapping = new ProductMapping(mappingParams)
  }

  run(input: stream$Readable, output: stream$Writable) {
    const productStream = this.parse(input, output)
    const { headerFields } = this.parserConfig

    if (headerFields)
      writeToSingleCsvFile(
        productStream,
        output,
        this.logger,
        headerFields,
        this.parserConfig.delimiter
      )
    else
      writeToZipFile(
        productStream,
        output,
        this.logger,
        this.parserConfig.delimiter
      )
  }

  parse(input: stream$Readable, output: stream$Writable) {
    this.logger.info('Starting conversion')
    let productCount = 0

    return (
      highland(input)
        .through(JSONStream.parse('*'))
        .flatMap((product: ProductProjection) =>
          highland(this._resolveReferences(product))
        )
        .doto(() => {
          productCount += 1
          this.logger.debug(`Resolved references of ${productCount} products`)
        })
        // prepare the product objects for csv format
        .map((product: ResolvedProductProjection) =>
          this._productMapping.run(product)
        )
        .flatten()
        .doto(() => {
          this.logger.debug(`Done with conversion of ${productCount} products`)
        })
        .stopOnError((err: Error) => {
          this.logger.error(err)
          output.emit('error', err)
        })
    )
  }

  _resolveReferences(product: ProductProjection): ResolvedProductProjection {
    // ReferenceTypes that need to be resolved:
    // **ProductType
    // **Categories [array]
    // **TaxCategory
    // **State
    // **CategoryOrderHints

    return Promise.all([
      this._resolveProductType(product.productType),
      this._resolveTaxCategory(product.taxCategory),
      this._resolveState(product.state),
      this._resolveCategories(product.categories),
      this._resolveCategoryOrderHints(product.categoryOrderHints),
    ]).then(
      ([productType, taxCategory, state, categories, categoryOrderHints]: Array<
        Object
      >): ResolvedProductProjection => ({
        ...product,
        ...productType,
        ...taxCategory,
        ...state,
        ...categories,
        ...categoryOrderHints,
      })
    )
  }

  _resolveProductType(productTypeReference: TypeReference): Object {
    if (!productTypeReference) return {}

    const productTypeService = this._createService('productTypes')
    const uri = productTypeService.byId(productTypeReference.id).build()
    return this.fetchReferences(uri).then(
      ({
        body,
      }: SuccessResult): {
        productType: ProductType,
      } => ({ productType: body })
    )
  }

  _resolveTaxCategory(taxCategoryReference: TypeReference): Object {
    if (!taxCategoryReference) return {}

    const taxCategoryService = this._createService('taxCategories')
    const uri = taxCategoryService.byId(taxCategoryReference.id).build()
    return this.fetchReferences(uri).then(
      ({
        body,
      }: SuccessResult): {
        taxCategory: TaxCategory,
      } => ({ taxCategory: body })
    )
  }

  _resolveState(stateReference: TypeReference): Object {
    if (!stateReference) return {}

    const stateService = this._createService('states')
    const uri = stateService.byId(stateReference.id).build()
    return this.fetchReferences(uri).then(
      ({
        body,
      }: SuccessResult): {
        state: State,
      } => ({ state: body })
    )
  }

  _resolveCategories(
    categoriesReference: Array<TypeReference>
  ): Array<Category> | {} {
    if (!categoriesReference || !categoriesReference.length) return {}

    const categoryIds = categoriesReference.map(
      (category: TypeReference): string => category.id
    )
    return this._getCategories(categoryIds).then(
      (categories: Array<Category>): { categories: Array<Category> } => {
        if (this.parserConfig.categoryBy !== 'namedPath') return { categories }
        return Promise.map(
          categories,
          (cat: Category): Array<Category> => this._resolveAncestors(cat)
        ).then(
          (categoriesWithParents: Array<Category>): Object => ({
            categories: categoriesWithParents,
          })
        )
      }
    )
  }

  _resolveCategoryOrderHints(categoryOrderHintsReference: Object): Object {
    if (
      !categoryOrderHintsReference ||
      !Object.keys(categoryOrderHintsReference).length
    )
      return { categoryOrderHints: undefined }

    const catIdentifier = this.parserConfig.categoryOrderHintBy
    const lang = this.parserConfig.language
    const categoryIds = Object.keys(categoryOrderHintsReference)
    return this._getCategories(categoryIds).then(
      (resolvedCategories: Array<Category>): Object => {
        const categoryOrderHints = {}
        resolvedCategories.forEach((category: Category) => {
          const catRef =
            catIdentifier === 'name'
              ? category[catIdentifier][lang]
              : category[catIdentifier]
          categoryOrderHints[catRef] = categoryOrderHintsReference[category.id]
        })
        return { categoryOrderHints }
      }
    )
  }

  _resolveAncestors(category: Category): Promise<Category> {
    const getParent = (cat: Category): Promise<Category> => {
      if (!cat.parent) return Promise.resolve(cat)

      return this._getCategories([cat.parent.id])
        .then(
          (resolvedCategory: Object): Promise<Category> =>
            getParent(resolvedCategory[0])
        )
        .then((parent: Object): Promise<Category> => ({ ...cat, parent }))
    }
    return getParent(category)
  }

  // This method decides if to get the categories from cache or API
  _getCategories(ids: Array<string>): Promise<Array<Object>> {
    const notCachedIds = []
    const cachedCategories = []
    ids.forEach((id: string) => {
      if (this.categoriesCache[id])
        cachedCategories.push(this.categoriesCache[id])
      else notCachedIds.push(id)
    })
    if (!notCachedIds.length) return Promise.resolve(cachedCategories)

    const predicate = `id in ("${notCachedIds.join('", "')}")`
    const categoriesService = this._createService('categories')
    const uri = categoriesService.where(predicate).build()
    return this.fetchReferences(uri).then(
      ({ body: { results } }: SuccessResult): Array<Category> => {
        results.forEach((result: Category) => {
          cachedCategories.push(result)
          this.categoriesCache[result.id] = result
        })
        return cachedCategories
      }
    )
  }

  _createService(serviceType: string): Object {
    const service = createRequestBuilder({
      projectKey: this.apiConfig.projectKey,
    })[serviceType]

    return service
  }
}

ProductJsonToCsv.prototype.fetchReferences = memoize(function _fetchReferences(
  uri: string
): Promise<SuccessResult> {
  const request: ClientRequest = {
    uri,
    method: 'GET',
  }
  if (this.accessToken)
    request.headers = {
      Authorization: `Bearer ${this.accessToken}`,
    }
  return this.client.execute(request)
})
