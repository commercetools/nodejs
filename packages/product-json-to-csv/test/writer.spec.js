import fs from 'fs'
import highland from 'highland'
import tmp from 'tmp'
import StreamTest from 'streamtest'
import unzip from 'unzip'
import * as writer from '../src/writer'

const streamTest = StreamTest.v2

describe('Writer', () => {
  let logger
  let sampleProducts
  beforeEach(() => {
    logger = { info: jest.fn() }
    sampleProducts = [
      {
        id: '12345ab-id',
        key: 'product-key',
        productType: 'product-type-1',
        'variant.id': 1,
        'variant.sku': 'A0E200000001YKI',
        'variant.images': 'image-url1',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
      },
      {
        'variant.id': 6,
        'variant.sku': 'A0E200001YKI123',
        'variant.images': 'https://image1.de;myImage.com',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
      },
      {
        id: '75345ab-id',
        key: 'product-key-2',
        productType: 'product-type-2',
        'variant.id': 1,
        'variant.sku': 'A0E200000001YKI',
        'variant.images': 'https://example.com/foobar/commer.jpg|3|4',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 089 WHT',
        color: 'white',
        'colorFreeDefinition.en': 'black-white',
        'colorFreeDefinition.de': 'schwarz-weiß',
        designer: 'michaelkors',
      },
      {
        'variant.id': 3,
        'variant.sku': 'A0E200001YKI123',
        'variant.images': 'https://example.com/foobar/commer234.jpg|3|3',
        'colorFreeDefinition.de': 'schwarz-weiß',
        designer: 'michaelkors',
      },
      {
        id: 'another5ab-id',
        key: 'product-key',
        productType: 'product-type-1',
        'variant.id': 1,
        'variant.sku': 'A0E200001YKI',
        addedAttr: '',
        anotherAddedAttr: '',
        article: 'sample 777 WHT',
        color: 'grune',
        'colorFreeDefinition.en': 'grey-white',
      },
      {
        'variant.id': 21,
        'variant.sku': 'PPP0001YKI123',
        'variant.images': 'https://eg993',
        'colorFreeDefinition.en': 'schwarz-weiß',
        color: 'blau',
      },
    ]
  })

  describe('::writeToSingleCsvFile', () => {
    test('write products to a single file with specified headers', done => {
      const sampleStream = highland(sampleProducts)
      const headers = [
        'id',
        'key',
        'productType',
        'variantId',
        'sku',
        'images',
        'anotherAddedAttr',
        'article',
        'designer',
        'color',
      ]
      const outputStream = streamTest.toText((error, actual) => {
        expect(error).toBeFalsy()
        expect(actual).toMatchSnapshot()
        done()
      })

      writer.writeToSingleCsvFile(sampleStream, outputStream, logger, headers)
    })

    test('write products to a single file with specified delimiter', done => {
      const sampleStream = highland(sampleProducts)
      const headers = [
        'id',
        'key',
        'productType',
        'variantId',
        'sku',
        'images',
        'anotherAddedAttr',
        'article',
        'designer',
        'color',
      ]
      const delimiter = ';'
      const outputStream = streamTest.toText((error, actual) => {
        expect(error).toBeFalsy()
        expect(actual).toMatchSnapshot()
        done()
      })

      writer.writeToSingleCsvFile(
        sampleStream,
        outputStream,
        logger,
        headers,
        delimiter
      )
    })

    test('log success info on completion', done => {
      const sampleStream = highland(sampleProducts)
      const headers = []
      const outputStream = streamTest.toText(() => {})
      outputStream.on('finish', () => {
        expect(logger.info).toBeCalledWith(
          expect.stringMatching(/written to CSV file/)
        )
        done()
      })

      writer.writeToSingleCsvFile(sampleStream, outputStream, logger, headers)
    })
  })

  describe('::writeToZipFile', () => {
    test('write products to multiple files based on productTypes', done => {
      const sampleStream = highland(sampleProducts)
      const verifyCsv1 = streamTest.toText((error, actual) => {
        expect(error).toBeFalsy()
        expect(actual).toMatchSnapshot()
      })
      const verifyCsv2 = streamTest.toText((error, actual) => {
        expect(error).toBeFalsy()
        expect(actual).toMatchSnapshot()
      })

      const tempFile = tmp.fileSync({ postfix: '.zip', keep: true })
      const output = tempFile.name
      const outputStream = fs.createWriteStream(output)

      // Extract data from zip file and test
      outputStream.on('finish', () => {
        fs.createReadStream(output)
          .pipe(unzip.Parse())
          .on('entry', entry => {
            if (entry.path === 'products/product-type-1.csv')
              entry.pipe(verifyCsv1)
            else if (entry.path === 'products/product-type-2.csv')
              entry.pipe(verifyCsv2)
          })
          .on('finish', () => {
            tempFile.removeCallback()
            done()
          })
      })

      writer.writeToZipFile(sampleStream, outputStream, logger)
    })

    test('log success info on completion', done => {
      const sampleStream = highland(sampleProducts)

      const tempFile = tmp.fileSync({ postfix: '.zip' })
      const output = tempFile.name
      const outputStream = fs.createWriteStream(output)

      outputStream.on('finish', () => {
        expect(logger.info).toBeCalledWith(
          expect.stringMatching(/written to ZIP file/)
        )
        tempFile.removeCallback()
        done()
      })

      writer.writeToZipFile(sampleStream, outputStream, logger)
    })
  })
})
