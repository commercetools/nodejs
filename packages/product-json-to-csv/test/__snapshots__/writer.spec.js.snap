// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Writer ::writeToSingleCsvFile write products to a single file with specified delimiter 1`] = `
"id;key;productType;variantId;sku;images;anotherAddedAttr;article;designer;color
\\"12345ab-id\\";\\"product-key\\";\\"product-type-1\\";1;\\"A0E200000001YKI\\";\\"image-url1\\";\\"\\";\\"sample 089 WHT\\";;\\"white\\"
;;;6;\\"A0E200001YKI123\\";\\"https://image1.de;myImage.com\\";;\\"sample 089 WHT\\";;\\"white\\"
\\"75345ab-id\\";\\"product-key-2\\";\\"product-type-2\\";1;\\"A0E200000001YKI\\";\\"https://example.com/foobar/commer.jpg|3|4\\";\\"\\";\\"sample 089 WHT\\";\\"michaelkors\\";\\"white\\"
;;;3;\\"A0E200001YKI123\\";\\"https://example.com/foobar/commer234.jpg|3|3\\";;;\\"michaelkors\\";
\\"another5ab-id\\";\\"product-key\\";\\"product-type-1\\";1;\\"A0E200001YKI\\";;\\"\\";\\"sample 777 WHT\\";;\\"grune\\"
;;;21;\\"PPP0001YKI123\\";\\"https://eg993\\";;;;\\"blau\\"
"
`;

exports[`Writer ::writeToSingleCsvFile write products to a single file with specified headers 1`] = `
"id,key,productType,variantId,sku,images,anotherAddedAttr,article,designer,color
\\"12345ab-id\\",\\"product-key\\",\\"product-type-1\\",1,\\"A0E200000001YKI\\",\\"image-url1\\",\\"\\",\\"sample 089 WHT\\",,\\"white\\"
,,,6,\\"A0E200001YKI123\\",\\"https://image1.de;myImage.com\\",,\\"sample 089 WHT\\",,\\"white\\"
\\"75345ab-id\\",\\"product-key-2\\",\\"product-type-2\\",1,\\"A0E200000001YKI\\",\\"https://example.com/foobar/commer.jpg|3|4\\",\\"\\",\\"sample 089 WHT\\",\\"michaelkors\\",\\"white\\"
,,,3,\\"A0E200001YKI123\\",\\"https://example.com/foobar/commer234.jpg|3|3\\",,,\\"michaelkors\\",
\\"another5ab-id\\",\\"product-key\\",\\"product-type-1\\",1,\\"A0E200001YKI\\",,\\"\\",\\"sample 777 WHT\\",,\\"grune\\"
,,,21,\\"PPP0001YKI123\\",\\"https://eg993\\",,,,\\"blau\\"
"
`;

exports[`Writer ::writeToZipFile write products to multiple files based on productTypes 1`] = `
"\\"id\\",\\"key\\",\\"productType\\",\\"variantId\\",\\"sku\\",\\"images\\",\\"addedAttr\\",\\"anotherAddedAttr\\",\\"article\\",\\"color\\",\\"colorFreeDefinition.en\\"
\\"12345ab-id\\",\\"product-key\\",\\"product-type-1\\",1,\\"A0E200000001YKI\\",\\"image-url1\\",\\"\\",\\"\\",\\"sample 089 WHT\\",\\"white\\",\\"black-white\\"
,,,6,\\"A0E200001YKI123\\",\\"https://image1.de;myImage.com\\",,,\\"sample 089 WHT\\",\\"white\\",\\"black-white\\"
\\"another5ab-id\\",\\"product-key\\",\\"product-type-1\\",1,\\"A0E200001YKI\\",,\\"\\",\\"\\",\\"sample 777 WHT\\",\\"grune\\",\\"grey-white\\"
,,,21,\\"PPP0001YKI123\\",\\"https://eg993\\",,,,\\"blau\\",\\"schwarz-weiß\\"
"
`;

exports[`Writer ::writeToZipFile write products to multiple files based on productTypes 2`] = `
"\\"id\\",\\"key\\",\\"productType\\",\\"variantId\\",\\"sku\\",\\"images\\",\\"addedAttr\\",\\"anotherAddedAttr\\",\\"article\\",\\"color\\",\\"colorFreeDefinition.en\\",\\"colorFreeDefinition.de\\",\\"designer\\"
\\"75345ab-id\\",\\"product-key-2\\",\\"product-type-2\\",1,\\"A0E200000001YKI\\",\\"https://example.com/foobar/commer.jpg|3|4\\",\\"\\",\\"\\",\\"sample 089 WHT\\",\\"white\\",\\"black-white\\",\\"schwarz-weiß\\",\\"michaelkors\\"
,,,3,\\"A0E200001YKI123\\",\\"https://example.com/foobar/commer234.jpg|3|3\\",,,,,,\\"schwarz-weiß\\",\\"michaelkors\\"
"
`;
